# Lesson 3 Assignment - Monty Hall Simulation
## Author - Shuyang Ren

### Instructions
Simulate the conditional probabilities of the distributions for winning a car vs. a goat for the Monty Hall Problem that includes the following:
1. the cases for switching or not switching
2. histograms and summary statistics (at least mean and variance) of the results for winning the car or a goat for both strategies: switching or not switching
3. comparison of the cases to determine the best strategy
4. recommended strategy



#### Note: 
Please clearly outline the results and your recommendation with markdown text. Be sure your discussion includes narrative for your figures and tables so it is clear to the reader the importance of each bit of evidence.

In [1]:
# Import libraries
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import random

In [2]:
successful_switches = [] # list of games where a contestant switches their pick ends up winning
successful_stays = [] # list of games where a contestant staying with their ininital pick ends up winning
    
def simulate_single_game(does_switch):
    # the winning door and initial door pick are randomlized
    winning_door = random.randint(1, 3)
    initial_door = random.randint(1, 3)
    
    # setting up the doors that Monte could possibly pick to reveal to the contestant
    monte_doors = [1, 2, 3]
    # remove the winning door and inital pick from options that monte can reveal
    monte_doors.remove(winning_door)
    if (initial_door in monte_doors):
        monte_doors.remove(initial_door)
    monte_door = random.choice(monte_doors)
    
    # the switch door is the door Monte didn't open and the contestant didn't pick
    switch_doors = [1,2,3]
    switch_doors.remove(monte_door)
    switch_doors.remove(initial_door)
    switch_door = switch_doors[0]

    # case of the contestant choosing to switching    
    if does_switch:
        return (switch_door == winning_door)
    # case for not switching
    return (initial_door == winning_door)

def simulate_game_trials(num_of_trials):
    for trial in range(num_of_trials):
        successful_switches.append(simulate_single_game(True))
        successful_stays.append(simulate_single_game(False))

In [3]:
# run the simulation for 100,000 trials
simulate_game_trials(100000)

In [4]:
# the percentage of games where switching won the contestant the game
np.mean(successful_switches)

0.66617

In [5]:
# the percentage of games where staying won the  contestant  the game
np.mean(successful_stays)

0.33416

In [6]:
# variance of winning games where contestant switched
np.var(successful_switches)

0.2223875311

In [7]:
# variance of winning games where contestant stayed
np.var(successful_stays)

0.22249709439999998

# Summary

From the results of 100,000 simulatin of the Monte Hall problem, the mean rate of winning games where the contestant switched their door choice is around 0.667 while that of games where the contestant stayed with their door choice is around 0.334. In either case the variance is about the same at around the value of 0.222. Therefore, we strongly recommand contestant to always opt for the  strategy of "switching" their door choice.